### DISCUSSION TITLE
'NeighborSampler' requires either 'pyg-lib' or 'torch-sparse'

### DISCUSSION THREAD

**Discussion by Rexedoziem** [pyg-team/pytorch_geometric#9143] [Q&A] [answered]
import torch_sparse
import torch_scatter
from torch_geometric.loader import NeighborLoader

train_loader = NeighborLoader(
    combined_data,
    num_neighbors=[10,10],
    batch_size=32,
    shuffle=True,
    input_nodes=('user', combined_data['user'].train_mask),
)
# Iterate over the train_loader and check for errors
for i, subgraph in enumerate(train_loader):
    print(f"Processing subgraph {i}")

After running the code, i had this error 'NeighborSampler' requires either 'pyg-lib' or 'torch-sparse'.
i am running it on windows 11pro using kaggle notebook for my analysis, i don't know where the problem is coming from as i've installed all necessary packages required?.

**Comment by akihironitta** [comment]
It looks like PyG can find niether pyg-lib nor torch-sparse. Can you share the output of `python -m pip list`? If you don't see them, you might want to try installing them again.

**Comment by Muyiiiii** [comment, marked as answer]
Just install pyg cluster sparse and so on before the torch_geometric .

**Comment by karthiksoman** [comment]
If you're still searching for a solution, here’s what worked for me:
```
conda create -n pyg_env python=3.12
conda activate pyg_env
pip install --no-binary=torch-sparse torch-sparse
pip install --no-binary=torch-scatter torch-scatter
pip install torch torchvision torchaudio
pip install torch-geometric
```

**Reply by Muyiiiii** [reply]
Good answer, I just don't know how to use pip install torch-sparse. I always install from wheel on this website https://[pytorch-geometric.com/whl/](https://pytorch-geometric.com/whl/)

**Comment by Srivathsan-Srinivas** [comment]
I had the same problem and it was fixed by doing the following in my mac:

1. Make sure `/Library/Frameworks/Python.framework/` is present with `Versions` subdirectory pointing to your preferred version.
```
(ve_3.10.13) Sri@Harnis-MacBook-Pro trial_4 % ls -l /Library/Frameworks/Python.framework/Versions 
total 0
drwxrwxr-x  11 root  admin  352 Nov  6 12:51 3.10
drwxrwxr-x  11 root  admin  352 Nov  5 15:45 3.11
lrwxr-xr-x   1 root  wheel    4 Nov  6 12:51 Current -> 3.10
(ve_3.10.13) Sri@Harnis-MacBook-Pro trial_4 % 
```
If, for any reason, you do not have the directory Python.framework (I did not have it a few days back), I installed python by downloading an installer from Python.org. `brew install python` did not create any subdir in Frameworks....for some reason.

2. For me, having torch version 2.4.0 combined with Python 3.10.x worked.
`pip3 install torch==2.4.0`.
3. Then, you can execute `pip install torch-geometric torch-sparse torch-scatter torch-cluster torch-spline-conv pyg-lib -f https://data.pyg.org/whl/torch-2.4.0+cpu.html`

note: do all the above in your preferred virtual-env (of 3.10.x).

**Reply by Anaraquelpengelly** [reply]
I had the same issue and I solved it by using a python3.9 virtualenv, and installing the following:

`pip install sentence-transformers` \n
`pip install torch==2.2.0`\n
`pip install torch_scatter torch_sparse torch_cluster torch_spline_conv -f https://data.pyg.org/whl/torch-2.2.0+cpu.html`
`pip install torch_geometric`
`pip install numpy==1.26.2`
